File: mobile-sdk/react-native-sdk/src/auth/AuthService.ts
import EncryptedStorage from 'react-native-encrypted-storage';
import { AuthTokens, User, LoginCredentials, ApiResponse } from '../types';
import { STORAGE_KEYS, API_ENDPOINTS } from '../utils/constants';
import { AuthenticationError, PyAirtableSDKError } from '../utils/errors';
import { withRetry } from '../utils/retry';

export class AuthService {
  private baseUrl: string;
  private apiKey: string;
  private tokens: AuthTokens | null = null;
  private user: User | null = null;
  private refreshPromise: Promise<void> | null = null;

  constructor(baseUrl: string, apiKey: string) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.initializeFromStorage();
  }

  /**
   * Initialize auth state from encrypted storage
   */
  private async initializeFromStorage(): Promise<void> {
    try {
      const [tokensStr, userStr] = await Promise.all([
        EncryptedStorage.getItem(STORAGE_KEYS.AUTH_TOKENS),
        EncryptedStorage.getItem(STORAGE_KEYS.USER_DATA),
      ]);

      if (tokensStr) {
        this.tokens = JSON.parse(tokensStr);
      }

      if (userStr) {
        this.user = JSON.parse(userStr);
      }

      // Check if tokens are expired and refresh if needed
      if (this.tokens && this.isTokenExpired()) {
        await this.refreshTokens();
      }
    } catch (error) {
      console.warn('Failed to initialize auth from storage:', error);
      await this.clearStoredAuth();
    }
  }

  /**
   * Login with email and password
   */
  async login(credentials: LoginCredentials): Promise<void> {
    try {
      const response = await this.makeRequest<{
        tokens: AuthTokens;
        user: User;
      }>(API_ENDPOINTS.AUTH_LOGIN, {
        method: 'POST',
        body: JSON.stringify(credentials),
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.success || !response.data) {
        throw new AuthenticationError(response.error || 'Login failed');
      }

      const { tokens, user } = response.data;
      await this.setTokens(tokens);
      await this.setUser(user);
    } catch (error) {
      throw error instanceof PyAirtableSDKError ? error : 
        new AuthenticationError('Login failed: ' + error.message);
    }
  }

  /**
   * Logout and clear all stored data
   */
  async logout(): Promise<void> {
    try {
      if (this.tokens) {
        await this.makeRequest(API_ENDPOINTS.AUTH_LOGOUT, {
          method: 'POST',
          headers: this.getAuthHeaders(),
        });
      }
    } catch (error) {
      console.warn('Logout request failed:', error);
    } finally {
      await this.clearStoredAuth();
    }
  }

  /**
   * Refresh access token using refresh token
   */
  async refreshTokens(): Promise<void> {
    // Prevent multiple simultaneous refresh attempts
    if (this.refreshPromise) {
      return this.refreshPromise;
    }

    this.refreshPromise = this.performTokenRefresh();
    
    try {
      await this.refreshPromise;
    } finally {
      this.refreshPromise = null;
    }
  }

  private async performTokenRefresh(): Promise<void> {
    if (!this.tokens?.refreshToken) {
      throw new AuthenticationError('No refresh token available');
    }

    try {
      const response = await this.makeRequest<{
        tokens: AuthTokens;
        user?: User;
      }>(API_ENDPOINTS.AUTH_REFRESH, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.tokens.refreshToken}`,
        },
      });

      if (!response.success || !response.data) {
        throw new AuthenticationError(response.error || 'Token refresh failed');
      }

      const { tokens, user } = response.data;
      await this.setTokens(tokens);
      
      if (user) {
        await this.setUser(user);
      }
    } catch (error) {
      await this.clearStoredAuth();
      throw error instanceof PyAirtableSDKError ? error : 
        new AuthenticationError('Token refresh failed: ' + error.message);
    }
  }

  /**
   * Get current user
   */
  getUser(): User | null {
    return this.user;
  }

  /**
   * Get current tokens
   */
  getTokens(): AuthTokens | null {
    return this.tokens;
  }

  /**
   * Check if user is authenticated
   */
  isAuthenticated(): boolean {
    return !!this.tokens && !this.isTokenExpired();
  }

  /**
   * Get authorization headers for API requests
   */
  getAuthHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'X-API-Key': this.apiKey,
    };

    if (this.tokens?.accessToken) {
      headers['Authorization'] = `Bearer ${this.tokens.accessToken}`;
    }

    return headers;
  }

  /**
   * Make authenticated API request with automatic token refresh
   */
  async makeAuthenticatedRequest<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    if (!this.isAuthenticated()) {
      throw new AuthenticationError('Not authenticated');
    }

    // Add auth headers
    const headers = {
      ...options.headers,
      ...this.getAuthHeaders(),
    };

    try {
      return await this.makeRequest<T>(endpoint, { ...options, headers });
    } catch (error) {
      // If token expired, try to refresh and retry once
      if (error instanceof AuthenticationError && this.tokens?.refreshToken) {
        try {
          await this.refreshTokens();
          
          // Retry with new token
          const newHeaders = {
            ...options.headers,
            ...this.getAuthHeaders(),
          };
          
          return await this.makeRequest<T>(endpoint, { ...options, headers: newHeaders });
        } catch (refreshError) {
          throw refreshError;
        }
      }
      
      throw error;
    }
  }

  /**
   * Make basic API request with retry logic
   */
  private async makeRequest<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    const url = `${this.baseUrl}${endpoint}`;
    
    return withRetry(
      async () => {
        const response = await fetch(url, {
          timeout: 30000,
          ...options,
          headers: {
            'Content-Type': 'application/json',
            ...options.headers,
          },
        });

        const data = await response.json();

        if (!response.ok) {
          const error = new PyAirtableSDKError(
            data.message || response.statusText,
            data.code,
            response.status,
            data
          );
          throw error;
        }

        return data as ApiResponse<T>;
      },
      {
        attempts: 3,
        delay: 1000,
      }
    );
  }

  /**
   * Check if current token is expired
   */
  private isTokenExpired(): boolean {
    if (!this.tokens) return true;
    
    // Add 5 minute buffer
    const expirationBuffer = 5 * 60 * 1000;
    return Date.now() >= (this.tokens.expiresAt - expirationBuffer);
  }

  /**
   * Store tokens in encrypted storage
   */
  private async setTokens(tokens: AuthTokens): Promise<void> {
    this.tokens = tokens;
    await EncryptedStorage.setItem(STORAGE_KEYS.AUTH_TOKENS, JSON.stringify(tokens));
  }

  /**
   * Store user data in encrypted storage
   */
  private async setUser(user: User): Promise<void> {
    this.user = user;
    await EncryptedStorage.setItem(STORAGE_KEYS.USER_DATA, JSON.stringify(user));
  }

  /**
   * Clear all stored authentication data
   */
  private async clearStoredAuth(): Promise<void> {
    this.tokens = null;
    this.user = null;
    
    await Promise.all([
      EncryptedStorage.removeItem(STORAGE_KEYS.AUTH_TOKENS),
      EncryptedStorage.removeItem(STORAGE_KEYS.USER_DATA),
    ]);
  }
}

File: mobile-sdk/flutter-sdk/lib/auth/auth_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;

import '../types/types.dart';
import '../utils/constants.dart';
import '../utils/retry.dart';

class AuthService {
  final String baseUrl;
  final String apiKey;
  final FlutterSecureStorage _secureStorage;
  
  AuthTokens? _tokens;
  User? _user;
  Completer<void>? _refreshCompleter;

  AuthService({
    required this.baseUrl,
    required this.apiKey,
  }) : _secureStorage = const FlutterSecureStorage();

  /// Initialize auth state from secure storage
  Future<void> initialize() async {
    try {
      final tokensJson = await _secureStorage.read(key: StorageKeys.authTokens);
      final userJson = await _secureStorage.read(key: StorageKeys.userData);

      if (tokensJson != null) {
        _tokens = AuthTokens.fromJson(jsonDecode(tokensJson));
      }

      if (userJson != null) {
        _user = User.fromJson(jsonDecode(userJson));
      }

      // Check if tokens are expired and refresh if needed
      if (_tokens != null && _tokens!.isExpired) {
        await refreshTokens();
      }
    } catch (error) {
      print('Failed to initialize auth from storage: $error');
      await _clearStoredAuth();
    }
  }

  /// Login with email and password
  Future<void> login(LoginCredentials credentials) async {
    try {
      final response = await _makeRequest<Map<String, dynamic>>(
        ApiEndpoints.authLogin,
        method: 'POST',
        body: credentials.toJson(),
      );

      if (!response.success || response.data == null) {
        throw AuthenticationException(response.error ?? 'Login failed');
      }

      final tokens = AuthTokens.fromJson(response.data!['tokens']);
      final user = User.fromJson(response.data!['user']);
      
      await _setTokens(tokens);
      await _setUser(user);
    } catch (error) {
      if (error is PyAirtableException) {
        rethrow;
      }
      throw AuthenticationException('Login failed: $error');
    }
  }

  /// Logout and clear all stored data
  Future<void> logout() async {
    try {
      if (_tokens != null) {
        await _makeRequest(
          ApiEndpoints.authLogout,
          method: 'POST',
          headers: _getAuthHeaders(),
        );
      }
    } catch (error) {
      print('Logout request failed: $error');
    } finally {
      await _clearStoredAuth();
    }
  }

  /// Refresh access token using refresh token
  Future<void> refreshTokens() async {
    // Prevent multiple simultaneous refresh attempts
    if (_refreshCompleter != null) {
      return _refreshCompleter!.future;
    }

    _refreshCompleter = Completer<void>();
    
    try {
      await _performTokenRefresh();
      _refreshCompleter!.complete();
    } catch (error) {
      _refreshCompleter!.completeError(error);
    } finally {
      _refreshCompleter = null;
    }
  }

  Future<void> _performTokenRefresh() async {
    if (_tokens?.refreshToken == null) {
      throw const AuthenticationException('No refresh token available');
    }

    try {
      final response = await _makeRequest<Map<String, dynamic>>(
        ApiEndpoints.authRefresh,
        method: 'POST',
        headers: {
          'Authorization': 'Bearer ${_tokens!.refreshToken}',
        },
      );

      if (!response.success || response.data == null) {
        throw AuthenticationException(response.error ?? 'Token refresh failed');
      }

      final tokens = AuthTokens.fromJson(response.data!['tokens']);
      await _setTokens(tokens);
      
      if (response.data!['user'] != null) {
        final user = User.fromJson(response.data!['user']);
        await _setUser(user);
      }
    } catch (error) {
      await _clearStoredAuth();
      if (error is PyAirtableException) {
        rethrow;
      }
      throw AuthenticationException('Token refresh failed: $error');
    }
  }

  /// Get current user
  User? get user => _user;

  /// Get current tokens
  AuthTokens? get tokens => _tokens;

  /// Check if user is authenticated
  bool get isAuthenticated => _tokens != null && !_tokens!.isExpired;

  /// Get authorization headers for API requests
  Map<String, String> _getAuthHeaders() {
    final headers = <String, String>{
      'X-API-Key': apiKey,
    };

    if (_tokens?.accessToken != null) {
      headers['Authorization'] = 'Bearer ${_tokens!.accessToken}';
    }

    return headers;
  }

  /// Make authenticated API request with automatic token refresh
  Future<ApiResponse<T>> makeAuthenticatedRequest<T>(
    String endpoint, {
    String method = 'GET',
    Map<String, dynamic>? body,
    Map<String, String>? headers,
    T Function(dynamic)? fromJson,
  }) async {
    if (!isAuthenticated) {
      throw const AuthenticationException('Not authenticated');
    }

    // Add auth headers
    final authHeaders = {
      ...?headers,
      ..._getAuthHeaders(),
    };

    try {
      return await _makeRequest<T>(
        endpoint,
        method: method,
        body: body,
        headers: authHeaders,
        fromJson: fromJson,
      );
    } catch (error) {
      // If token expired, try to refresh and retry once
      if (error is AuthenticationException && _tokens?.refreshToken != null) {
        try {
          await refreshTokens();
          
          // Retry with new token
          final newAuthHeaders = {
            ...?headers,
            ..._getAuthHeaders(),
          };
          
          return await _makeRequest<T>(
            endpoint,
            method: method,
            body: body,
            headers: newAuthHeaders,
            fromJson: fromJson,
          );
        } catch (refreshError) {
          rethrow;
        }
      }
      
      rethrow;
    }
  }

  /// Make basic API request with retry logic
  Future<ApiResponse<T>> _makeRequest<T>(
    String endpoint, {
    String method = 'GET',
    Map<String, dynamic>? body,
    Map<String, String>? headers,
    T Function(dynamic)? fromJson,
  }) async {
    final url = Uri.parse('$baseUrl$endpoint');
    
    return withRetry(
      () async {
        final request = http.Request(method, url);
        
        request.headers.addAll({
          'Content-Type': 'application/json',
          ...?headers,
        });

        if (body != null) {
          request.body = jsonEncode(body);
        }

        final streamedResponse = await request.send();
        final response = await http.Response.fromStream(streamedResponse);
        
        final responseData = jsonDecode(response.body);

        if (response.statusCode < 200 || response.statusCode >= 300) {
          throw _createErrorFromResponse(response, responseData);
        }

        return ApiResponse<T>.fromJson(responseData, fromJson);
      },
      const RetryOptions(
        attempts: 3,
        delay: 1000,
        shouldRetry: isRetryableError,
      ),
    );
  }

  /// Create error from HTTP response
  PyAirtableException _createErrorFromResponse(
    http.Response response,
    dynamic data,
  ) {
    final statusCode = response.statusCode;
    final message = data is Map<String, dynamic> ? data['message'] : response.reasonPhrase;
    
    switch (statusCode) {
      case HttpStatus.badRequest:
        return ValidationException(message ?? 'Bad request', details: data);
      case HttpStatus.unauthorized:
        return AuthenticationException(message ?? 'Unauthorized');
      case HttpStatus.forbidden:
        return PermissionException(message ?? 'Forbidden');
      case HttpStatus.notFound:
        return NotFoundException(message ?? 'Not found');
      case HttpStatus.conflict:
        return ValidationException(message ?? 'Conflict', details: data);
      case HttpStatus.unprocessableEntity:
        return ValidationException(message ?? 'Validation failed', details: data);
      case HttpStatus.tooManyRequests:
        final retryAfter = data is Map<String, dynamic> ? data['retryAfter'] : null;
        return RateLimitException(retryAfter: retryAfter);
      case HttpStatus.internalServerError:
      case HttpStatus.badGateway:
      case HttpStatus.serviceUnavailable:
        return ServerException(message ?? 'Server error', statusCode: statusCode);
      default:
        return PyAirtableException(
          message ?? 'Unknown error',
          code: 'UNKNOWN_ERROR',
          statusCode: statusCode,
          details: data,
        );
    }
  }

  /// Store tokens in secure storage
  Future<void> _setTokens(AuthTokens tokens) async {
    _tokens = tokens;
    await _secureStorage.write(
      key: StorageKeys.authTokens,
      value: jsonEncode(tokens.toJson()),
    );
  }

  /// Store user data in secure storage
  Future<void> _setUser(User user) async {
    _user = user;
    await _secureStorage.write(
      key: StorageKeys.userData,
      value: jsonEncode(user.toJson()),
    );
  }

  /// Clear all stored authentication data
  Future<void> _clearStoredAuth() async {
    _tokens = null;
    _user = null;
    
    await Future.wait([
      _secureStorage.delete(key: StorageKeys.authTokens),
      _secureStorage.delete(key: StorageKeys.userData),
    ]);
  }
}